### ISSUE TITLE
Tunnelblick 3.8.x on MacOS: Any setting for "script-security" in configuration still shows the warning

### ISSUE THREAD

**Issue by mackonsti** [Tunnelblick/Tunnelblick#666]
**Describe the bug**
I noticed that when I add/launch an `.ovpn` configuration file to connect to either my router or NAS, Tunnelblick recent macOS versions showed a warning:
`NOTE: the current --script-security setting may allow this configuration to call user-defined scripts`

Despite adding `script-security` in these `.ovpn` configurations with values 0 or 1, the warning keeps showing. This may be intentional, but with value 0 the warning should not be seen on the log window.

**To Reproduce**
1. Install Tunnelblick 3.x on macOS;
2. Add a working configuration file, with`script-security` option inside set to 0 or 1;
3. Connect to that specific OpenVPN server;
4. This warning appears no matter the value set in the configuration file.

**Expected behavior**
I would personally expect _at least_ with value 0 to not show this warning; I have no idea if this is a fixed warning in the code of Tunnelblick, regardless of the setting in `script-security` of each configuration file. Even if it's not something important or critical, getting this warning prompts to enthusiasts like me to some possible improvement in the whole set up.

One would think that it follows what is found here: https://community.openvpn.net/openvpn/wiki/Openvpn24ManPage

**Screenshots**
This error is straight-forward in the log output window of Tunnelblick on macOS, highlighted in light-blue colour (when in dark mode):
`NOTE: the current --script-security setting may allow this configuration to call user-defined scripts`

**Additional context**
I do not know if I need to **add** somewhere the option `--script-security 0` in some default configuration file of the _actual_ Tunnelblick application itself; I just launch it on macOS from Applications folder, without messing with hidden configuration files; the Preferences window has no mention, nor do I know where to find the "default" client settings. But I would expect that configurations **override** this anyway.

**Diagnostic information**
Please see attached (after I removed my sensitive details):

[Log_Sanitized.txt](https://github.com/Tunnelblick/Tunnelblick/files/5922033/Log_Sanitized.txt)

Thank you for replying your insight on this issue, despite not being of critical or major importance.

**Comment by jkbullard** [comment]
Tunnelblick uses several of its own scripts to provide a lot of it's functionality when a VPN is connecting and disconnecting (see [Using Scripts](https://tunnelblick.net/cUsingScripts.html) for details). Implementing DNS changes requested by OpenVPN is the most common function they are used for, but there are others. So when you are using Tunnelblick's scripts, Tunnelblick adds a "--script-security 2" option to the command line in such a way that it overrides what is in the OpenVPN configuration file.

If you are using Tunnelblick for DNS changes, etc., then there is no way around that.

If you are not using Tunnelblick for DNS changes, etc., then set "Set DNS/WINS" to "Do not set nameserver" and Tunnelblick won't add "--script security 2" and the "script-security 2" in your configuration file should be in effect.

**Comment by jkbullard** [comment]
@mackonsti - Because your OpenVPN server is pushing "dhcp-option DNS…", you need the Tunnelblick "Set nameserver" setting, so Tunnelblick needs to use its scripts, so Tunnelblick will add the "--script-security 2" option when it starts OpenVPN.

**Comment by mackonsti** [comment]
Hello @jkbullard I appreciate your reply and the time you spent for me, despite being a trivial issue/question for you experienced guys :-D

Please know that I am not using any scripts myself, just an honest-to-God connection to my home's router or NAS, as I have both running as backup (and both show this warning). So I am just launching a fresh installation of Tunnelblick to macOS (Catalina in my case) thus I let it add its own options like `--script-security 2` to its startup procedure = I did see this when I read the log after posting here.

I will try seeing if setting "Set DNS/WINS" to "Do not set nameserver" option with my untouched script has any effect to this warning (and usability) before trying to see if my router _does_ need the dhcp-option DNS [...] setting for a connection to be established. I guess trial and error (and studying is needed) after following the [OpenWRT OpenVPN guide](https://openwrt.org/docs/guide-user/services/vpn/openvpn/server) here.

Kindly allow me a few days to do my tests and report back, for the sake of other Tunnelblick users, before I ask for your input, insight or closing this ticket. Many thanks, stay safe.

**Comment by jkbullard** [comment]
@mackonsti - Describing the results of your experiments is fine, but they won't necessarily apply to other users' configurations.

Please understand that you do not need to be so concerned about this warning. Although OpenVPN issues the warning, Tunnelblick makes sure that scripts that are not authorized by a computer administrator will not be run, so it doesn't really matter.

However, to remove this warning, you could do the following three things:

 1. Set your Mac to always use 8.8.8.8 and 8.8.4.4 as DNS addresses. (You can do that in System Preferences >> Network >> Advanced >> DNS.)

 2. Set "Set DNS/WINS" to "Do not set nameserver" in the "Settings" tab of Tunnelblick's "VPN Details" window.

 3. Modify the OpenVPN configuration file on your router to not "push" the "redirect-gateway def1 bypass-dhcp", "dhcp-option DNS 8.8.8.8", and "dhcp-option DNS 8.8.4.4" options. You should be able to find lines like the following that do that:
```
push "redirect-gateway def1 bypass-dhcp"
push "dhcp-option DNS 8.8.8.8"
push "dhcp-option DNS 8.8.4.4"
```
It is possible that the lines will not be together, or that they will be in a different order.

The effect of these three things will be that your computer will always use Google's DNS servers, instead of only using them when the VPN is active. Currently your computer gets its DNS settings from the results of a DHCP request. Those settings will vary, depending on what network your computer is connected to, but on the network you were using when you produced the diagnostic info that you posted, DNS is routed to 192.68.1.1, which is very common, and which is almost certainly the router your computer was connecting to the Internet through. To know what DNS was actually used by the router, you'd have to examine what the router itself does.

<hr>

Totally separately, you should probably put a check in the "Disable IPv6 unless the VPN server is accessed using IPv6" setting. Without that, your computer risks leaking information about your computer's public IP address even though you are using a VPN.

If you are using OpenVPN 2.5 on both the server and in Tunnelblick, you might be able to skip that and instead use the new "block-ipv6" OpenVPN option to block IPv6 traffic.

**Comment by mackonsti** [comment]
Agreed @jkbullard you caught me before doing anything (new). Thank you for your valuable input. It seems you are on top of things, your insight is very useful. Indeed this warning in question can be ignored.

I have already disabled IPv6 on macOS Catalina via the needed Terminal command `sudo networksetup -setv6off Ethernet` for my cable "Ethernet" named connection; your tip is however very useful as a good reminder in case IPv6 was left active, by default.

In view of the above, and for this casual use (as seen from my Log.txt shared in my first post) do we need to **remove** the setting `script-security X` from the client configurations, therefore, if using a typical OpenVPN Server setup that has `dhcp-option` set? I imagine it won't hurt leaving it.

I hope this thread may help others if found via a search engine like Google. I think we can close this issue ticket. Warm thanks from France. Stay safe.

**Comment by jkbullard** [comment]
@mackonsti - Leaving `script-security 0` in client configurations is fine, since it will be ignored if Set DNS/WINS is set to any of the "Set nameserver" settings. But I wouldn't recommend leaving other levels of script security in configuration files, because they would not be ignored if the user, perhaps accidentally, change to "Do not set nameserver".

**Comment by mackonsti** [comment]
Good day @jkbullard thanks for the recommendation; but for this to happen (i.e. _not_ be ignored as you warn) we do need the OpenVPN server to _still_ not set/push "dhcp-option" as described earlier, correct? Would a value of `script-security 1` instead be advised, therefore as a kind of balanced value?

**Comment by jkbullard** [comment]
@mackonsti  - The short answers are no, and yes.

If the server "pushes" DNS settings, they might be ignored by OpenVPN, or they might trigger an error. To make sure they can't trigger an error, don't "push" them.

Thank's for suggesting the use of `script-security 1`. You're correct that the OpenVPN configuration file should contain `script-security 1`. `script-security 0` will probably prevent the VPN from functioning properly. My earlier instructions to use `script-security 0` were wrong and I have edited my comment to fix that. I misremembered what that the settings do and didn't bother to look it up until you suggested `script-security 1`.
